// @serenity-js/web/lib/scripts/dragAndDrop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dragAndDrop = dragAndDrop;
/* c8 ignore start */
/**
 * Simulates HTML5 drag and drop events.
 *
 * This script is a workaround for Selenium WebDriver's inability to properly
 * trigger HTML5 drag events using the Actions API. It manually dispatches
 * the required drag events (dragstart, dragenter, dragover, drop, dragend)
 * to simulate a complete drag and drop operation.
 *
 * @param source - The element to drag
 * @param target - The element to drop onto
 */
function dragAndDrop(source, target) {
    function createDragEvent(type, dataTransfer) {
        return new DragEvent(type, {
            bubbles: true,
            cancelable: true,
            dataTransfer,
        });
    }
    function createDataTransfer() {
        try {
            return new DataTransfer();
        }
        catch {
            // Fallback for older browsers
            const event = document.createEvent('CustomEvent');
            event.initCustomEvent('', false, false, null);
            event.dataTransfer = {
                data: {},
                setData(format, data) {
                    this.data[format] = data;
                },
                getData(format) {
                    return this.data[format];
                },
                clearData(format) {
                    if (format) {
                        delete this.data[format];
                    }
                    else {
                        this.data = {};
                    }
                },
                setDragImage() {
                    // no-op
                },
                dropEffect: 'none',
                effectAllowed: 'all',
                files: [],
                items: [],
                types: [],
            };
            return event.dataTransfer;
        }
    }
    // Create a DataTransfer object to share between events
    const dataTransfer = createDataTransfer();
    // Dispatch dragstart on source
    source.dispatchEvent(createDragEvent('dragstart', dataTransfer));
    // Dispatch dragenter on target
    target.dispatchEvent(createDragEvent('dragenter', dataTransfer));
    // Dispatch dragover on target (required for drop to work)
    const dragOverEvent = createDragEvent('dragover', dataTransfer);
    target.dispatchEvent(dragOverEvent);
    // Dispatch drop on target
    target.dispatchEvent(createDragEvent('drop', dataTransfer));
    // Dispatch dragend on source
    source.dispatchEvent(createDragEvent('dragend', dataTransfer));
}
/* c8 ignore stop */
