// @serenity-js/serenity-bdd/lib/cli/model/GAV.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GAV = void 0;
const io_1 = require("@serenity-js/core/lib/io");
const tiny_types_1 = require("tiny-types");
/**
 * @package
 */
class GAV extends tiny_types_1.TinyType {
    constructor(groupId, artifactId, version, extension = 'jar', classifier) {
        super();
        this.groupId = groupId;
        this.artifactId = artifactId;
        this.version = version;
        this.extension = extension;
        this.classifier = classifier;
        tiny_types_1.ensure('groupId', groupId, tiny_types_1.isDefined(), tiny_types_1.matches(/^[a-z][a-z0-9_-]+(?:\.[a-z0-9_-]+)+[0-9a-z_-]$/));
        tiny_types_1.ensure('artifactId', artifactId, tiny_types_1.isDefined(), tiny_types_1.matches(/^[a-z0-9_-]+$/));
        tiny_types_1.ensure('version', version, tiny_types_1.isDefined(), tiny_types_1.matches(/^(?:\d+\.?){3}$/));
        tiny_types_1.ensure('extension', extension, tiny_types_1.isDefined(), tiny_types_1.matches(/^[a-z]+$/));
    }
    static fromString(gav) {
        const parts = gav.split(':');
        tiny_types_1.ensure('GAV segments', parts, tiny_types_1.property('length', tiny_types_1.isInteger(), tiny_types_1.isInRange(3, 5)));
        switch (parts.length) {
            case 5: return new GAV(parts[0], parts[1], parts[4], parts[2], parts[3]);
            case 4: return new GAV(parts[0], parts[1], parts[3], parts[2]);
            default: return new GAV(parts[0], parts[1], parts[2]);
        }
    }
    toPath() {
        const name = [
            this.artifactId,
            this.version,
            this.classifier,
        ].filter(s => !!s).join('-');
        return new io_1.Path(`${name}.${this.extension}`);
    }
}
exports.GAV = GAV;
