// @serenity-js/serenity-bdd/lib/cli/stage/ProgressReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressReporter = void 0;
const events_1 = require("@serenity-js/core/lib/events");
const ProgressBar = require("progress");
const model_1 = require("../model");
/**
 * @desc
 *  Notifies the user of the progress of the Serenity BDD CLI jar download.
 *
 * @package
 */
class ProgressReporter {
    constructor(printer, stage = null, progressBar = null) {
        this.printer = printer;
        this.stage = stage;
        this.progressBar = progressBar;
    }
    assignedTo(stage) {
        return new ProgressReporter(this.printer, stage, this.progressBar);
    }
    notifyOf(event) {
        if (event instanceof events_1.ArtifactGenerated && event.artifact instanceof model_1.DownloadProgressReport) {
            const { downloadedBytes, totalBytes } = event.artifact.map(report => report);
            if (!this.progressBar) {
                this.progressBar = new ProgressBar(`[:bar] Downloaded :percent at :ratebps. We should be done in about :etas`, {
                    complete: '=',
                    incomplete: '-',
                    width: 20,
                    total: totalBytes,
                    stream: this.printer.out,
                });
            }
            this.progressBar.tick(downloadedBytes);
        }
    }
}
exports.ProgressReporter = ProgressReporter;
