// @serenity-js/serenity-bdd/lib/cli/screenplay/abilities/UseFileSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UseFileSystem = void 0;
const io_1 = require("@serenity-js/core/lib/io");
/**
 * @package
 */
class UseFileSystem {
    constructor(fileSystem) {
        this.fileSystem = fileSystem;
    }
    static at(root) {
        return new UseFileSystem(new io_1.FileSystem(root));
    }
    static as(actor) {
        return actor.abilityTo(UseFileSystem);
    }
    createReadStream(relativePathToFile) {
        return this.fileSystem.createReadStream(relativePathToFile);
    }
    createWriteStreamTo(relativePathToFile) {
        return this.fileSystem.createWriteStreamTo(relativePathToFile);
    }
    createDirectory(relativePathToDirectory) {
        return this.fileSystem.ensureDirectoryExistsAt(relativePathToDirectory);
    }
    remove(relativePathToFileOrDirectory) {
        return this.fileSystem.remove(relativePathToFileOrDirectory);
    }
    rename(source, destination) {
        return this.fileSystem.rename(source, destination);
    }
    attributesOf(relativePathToFileOrDirectory) {
        return this.fileSystem.stat(relativePathToFileOrDirectory);
    }
}
exports.UseFileSystem = UseFileSystem;
