// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/SerenityBDDReportContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SerenityBDDReportContext = void 0;
/**
 * @package
 */
class SerenityBDDReportContext {
    constructor() {
        this.report = {};
        this.steps = new Map();
        this.currentActivityId = null;
    }
    with(fn) {
        return fn(this);
    }
    build() {
        const eraseDuplicateExceptionReportFromParentSteps = (current) => {
            if (current.parentActivityId) {
                const parent = this.steps.get(current.parentActivityId.value);
                delete parent.step.exception;
                return eraseDuplicateExceptionReportFromParentSteps(parent);
            }
        };
        const testSteps = [];
        this.steps.forEach(current => {
            if (current.parentActivityId) {
                this.steps.get(current.parentActivityId.value).step.children.push(current.step);
                if (current.step.exception) {
                    eraseDuplicateExceptionReportFromParentSteps(current);
                }
            }
            else {
                testSteps.push(current.step);
            }
        });
        return {
            ...this.report,
            testSteps,
        }; // tslint:disable-line:no-object-literal-type-assertion
    }
}
exports.SerenityBDDReportContext = SerenityBDDReportContext;
