// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/transformations/activityRelatedArtifact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activityRelatedArtifact = void 0;
const io_1 = require("@serenity-js/core/lib/io");
const model_1 = require("@serenity-js/core/lib/model");
const crypto_1 = require("crypto");
const tiny_types_1 = require("tiny-types");
const util_1 = require("util");
/**
 * @package
 */
function activityRelatedArtifact(activityId, name, artifact, timestamp) {
    const differ = new io_1.AssertionReportDiffer({
        expected: line => `+ ${line}`,
        actual: line => `- ${line}`,
        matching: line => `  ${line}`,
    });
    return (report) => tiny_types_1.match(artifact)
        .when(model_1.HTTPRequestResponse, _ => report.with(httpRequestResponse(activityId, artifact.map(data => data), timestamp)))
        .when(model_1.TextData, _ => report.with(arbitraryData(activityId, name, artifact.map(artifactContents => artifactContents.data), timestamp)))
        .when(model_1.LogEntry, _ => report.with(arbitraryData(activityId, name, artifact.map(artifactContents => artifactContents.data), timestamp)))
        .when(model_1.AssertionReport, _ => report.with(arbitraryData(activityId, name, artifact.map(value => differ.diff(value.expected, value.actual)), timestamp)))
        .when(model_1.JSONData, _ => report.with(arbitraryData(activityId, name, artifact.map(value => JSON.stringify(value, null, 4)), timestamp)))
        .else(_ => report);
}
exports.activityRelatedArtifact = activityRelatedArtifact;
function httpRequestResponse(activityId, requestResponse, timestamp) {
    function mapToString(dictionary) {
        return Object.keys(dictionary).map(key => `${key}: ${dictionary[key]}`).join('\n');
    }
    function bodyToString(data) {
        if (data === null || data === undefined) {
            return '';
        }
        if (typeof data === 'string') {
            return data;
        }
        if (typeof data === 'object') {
            return JSON.stringify(data, null, 4);
        }
        return util_1.inspect(data);
    }
    return (context) => {
        if (context.steps.has(activityId.value)) {
            context.steps.get(activityId.value).step.restQuery = {
                method: requestResponse.request.method.toUpperCase(),
                path: requestResponse.request.url,
                content: bodyToString(requestResponse.request.data),
                contentType: requestResponse.request.headers['Content-Type'] || '',
                requestHeaders: mapToString(requestResponse.request.headers) || '',
                requestCookies: requestResponse.request.headers.Cookie || '',
                statusCode: requestResponse.response.status,
                responseHeaders: mapToString(requestResponse.response.headers) || '',
                responseCookies: requestResponse.response.headers.Cookie || '',
                responseBody: bodyToString(requestResponse.response.data),
            };
        }
        return context;
    };
}
function arbitraryData(activityId, name, contents, timestamp) {
    return (context) => {
        if (context.steps.has(activityId.value)) {
            const id = crypto_1.createHash('sha1')
                .update(name.value)
                .update(contents)
                .update(`${timestamp.toMillisecondTimestamp()}`)
                .digest('hex');
            context.steps.get(activityId.value).step.reportData.push({
                id: `report-data-${id}`,
                isEvidence: false,
                path: '',
                title: name.value,
                contents,
            });
        }
        return context;
    };
}
