// @serenity-js/serenity-bdd/lib/cli/screenplay/interactions/Spawn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Spawn = void 0;
const core_1 = require("@serenity-js/core");
const io_1 = require("@serenity-js/core/lib/io");
const child_process_1 = require("child_process");
const model_1 = require("../../model");
/**
 * @package
 */
class Spawn extends core_1.Interaction {
    constructor(pathToExecutable, args) {
        super();
        this.pathToExecutable = pathToExecutable;
        this.args = args;
    }
    static the(pathToExecutable, ...args) {
        return new Spawn(pathToExecutable, args);
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions & CollectsArtifacts} actor
     * @returns {Promise<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     * @see {@link @serenity-js/core/lib/screenplay/actor~CollectsArtifacts}
     */
    performAs(actor) {
        return actor.answer(this.pathToExecutable)
            .then(pathToExecutable => new Promise((resolve, reject) => {
            actor.collect(model_1.Notification.fromJSON({ message: `Spawning: ${pathToExecutable.value} ${this.args.join(' ')}` }));
            const spawned = child_process_1.spawn(pathToExecutable.value, this.args, { stdio: [process.stdin, process.stdout, process.stderr] });
            spawned.once('error', (error) => actor.collect(model_1.Complaint.fromJSON({ description: `Invoking Serenity BDD CLI has failed`, message: error.message, stack: error.stack })));
            spawned.once('exit', (exitCode) => exitCode !== 0
                ? reject(new model_1.ExecutionError(`The following process exited with ${exitCode}: ${pathToExecutable.value} ${this.args.join(' ')}`))
                : resolve(void 0));
        }));
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor executes ${this.pathToExecutable} with ${this.args}`;
    }
}
exports.Spawn = Spawn;
